packages/playwright-core/src/server/screenshotter.ts

```diff
@@ -107,50 +116,92 @@ export class Screenshotter {
-
-    const restoreBlinkingCaret = async () => {
-      await Promise.all(this._page.frames().map(async frame => {
-        frame.nonStallingEvaluateInExistingContext('window.__cleanupScreenshot && window.__cleanupScreenshot()', false, 'utility').catch(() => {});
-      }));
-    };
+  async _preparePageForScreenshot(progress: Progress, disableAnimations: boolean) {
     await Promise.all(this._page.frames().map(async frame => {
-      await frame.nonStallingEvaluateInExistingContext((function() {
+      await frame.nonStallingEvaluateInExistingContext('(' + (function(disableAnimations: boolean) {
         const styleTag = document.createElement('style');
         styleTag.textContent = `
-          * { caret-color: transparent !important; }
-          * > * { caret-color: transparent !important; }
-          * > * > * { caret-color: transparent !important; }
-          * > * > * > * { caret-color: transparent !important; }
+          *,
+          * > *,
+          * > * > *,
+          * > * > * > *,
           * > * > * > * > * { caret-color: transparent !important; }
         `;
         document.documentElement.append(styleTag);
+        const infiniteAnimationsToResume: Animation[] = [];
+        const cleanupCallbacks: (() => void)[] = [];
+
+        if (disableAnimations) {
+          const collectRoots = (root: Document | ShadowRoot, roots: (Document|ShadowRoot)[] = []): (Document|ShadowRoot)[] => {
+            roots.push(root);
+            const walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
+            do {
+              const node = walker.currentNode;
+              const shadowRoot = node instanceof Element ? node.shadowRoot : null;
+              if (shadowRoot)
+                collectRoots(shadowRoot, roots);
+            } while (walker.nextNode());
+            return roots;
+          };
+          const handleAnimations = (root: Document|ShadowRoot): void => {
+            for (const animation of root.getAnimations()) {
+              try {
+                animation.finish();
+              } catch (e) {
+                animation.cancel();
+                infiniteAnimationsToResume.push(animation);
+              }
+            }
+          };
+          for (const root of collectRoots(document)) {
+            const handleRootAnimations: (() => void) = handleAnimations.bind(null, root);
+            handleRootAnimations();
+            root.addEventListener('transitionrun', handleRootAnimations);
+            root.addEventListener('animationstart', handleRootAnimations);
+            cleanupCallbacks.push(() => {
+              root.removeEventListener('transitionrun', handleRootAnimations);
+              root.removeEventListener('animationstart', handleRootAnimations);
+            });
+          }
+        }
+
         window.__cleanupScreenshot = () => {
           styleTag.remove();
+          for (const animation of infiniteAnimationsToResume)
+            animation.play();
```

Comment:
Let's try/catch every call here, to ensure that we finish other cleanups when `play()` throws?

Reply:
done.